File: tv-fork-patches/patched/smarttubetv/src/main/java/com/liskovsoft/smartyoutubetv2/tv/ui/litetube/LitetubeActivationActivity.java
package com.liskovsoft.smartyoutubetv2.tv.ui.litetube;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.Gravity;
import android.view.KeyEvent;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.liskovsoft.sharedutils.mylogger.Log;
import com.liskovsoft.smartyoutubetv2.common.litetube.LitetubeApi;
import com.liskovsoft.smartyoutubetv2.common.litetube.LitetubePrefs;
import com.liskovsoft.smartyoutubetv2.tv.ui.main.SplashActivity;

import org.json.JSONObject;

import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * First-launch activation screen for Litetube TV.
 */
public class LitetubeActivationActivity extends Activity {

    private static final String TAG = "LitetubeActivation";
    private final AtomicBoolean stopFlag = new AtomicBoolean(false);
    private final java.util.concurrent.ExecutorService ioExecutor =
            Executors.newSingleThreadExecutor(r -> new Thread(r, "litetube-activation"));

    private TextView codeView;
    private TextView urlView;
    private TextView statusView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setGravity(Gravity.CENTER);
        root.setPadding(48, 48, 48, 48);
        root.setBackgroundColor(0xFF0A0A14);

        TextView title = new TextView(this);
        title.setText("Активация Litetube");
        title.setTextSize(28f);
        title.setTextColor(0xFFFFFFFF);
        title.setGravity(Gravity.CENTER);
        title.setPadding(0, 32, 0, 48);
        root.addView(title);

        codeView = new TextView(this);
        codeView.setTextSize(72f);
        codeView.setTextColor(0xFF7C5CFC);
        codeView.setGravity(Gravity.CENTER);
        codeView.setPadding(0, 0, 0, 16);
        root.addView(codeView);

        urlView = new TextView(this);
        urlView.setTextSize(18f);
        urlView.setTextColor(0xFF8888AA);
        urlView.setGravity(Gravity.CENTER);
        urlView.setPadding(0, 0, 0, 32);
        root.addView(urlView);

        statusView = new TextView(this);
        statusView.setTextSize(16f);
        statusView.setTextColor(0xFFAAAAFF);
        statusView.setGravity(Gravity.CENTER);
        root.addView(statusView);

        setContentView(root);

        LitetubePrefs prefs = LitetubePrefs.instance(this);
        String api = prefs.getLitetubeApiBase();
        Log.d(TAG, "bootstrapping activation against " + api);

        ioExecutor.submit(() -> runActivationLoop(prefs, api));
    }

    private void runActivationLoop(LitetubePrefs prefs, String apiBase) {
        try {
            JSONObject start = LitetubeApi.startDevice(apiBase);
            if (start == null) {
                showFailure("Не удалось связаться с сервером активации. Проверьте интернет.");
                return;
            }
            String code = start.optString("code");
            if (code == null || code.length() != 6) {
                showFailure("Сервер вернул неожиданный ответ.");
                return;
            }
            String qrUrl = start.optString("qr_url");
            if (qrUrl == null || qrUrl.isEmpty()) {
                qrUrl = apiBase + "/activate?code=" + code;
            }
            prefs.setActivationCode(code);
            renderCode(code, qrUrl);
            showStatus("Ожидаем подтверждения на телефоне...");

            while (!stopFlag.get()) {
                JSONObject poll = LitetubeApi.pollDevice(apiBase, code);
                String status = poll != null ? poll.optString("status") : null;
                if ("claimed".equals(status)) {
                    String jwt = poll.optString("jwt");
                    if (jwt == null || jwt.isEmpty()) {
                        showFailure("Сервер не выдал токен. Попробуйте ещё раз.");
                        return;
                    }
                    prefs.setJwt(jwt);
                    prefs.clearActivationCode();
                    showStatus("Готово. Запускаем плеер...");
                    try { Thread.sleep(700); } catch (InterruptedException ignored) {}
                    proceedToSplash();
                    return;
                } else if ("expired".equals(status)) {
                    showFailure("Код истёк. Перезапустите приложение.");
                    return;
                }
                try { Thread.sleep(TimeUnit.SECONDS.toMillis(2)); } catch (InterruptedException ignored) {}
            }
        } catch (Throwable t) {
            Log.e(TAG, "activation loop crashed: " + t.getMessage(), t);
            showFailure("Ошибка: " + (t.getMessage() != null ? t.getMessage() : t.getClass().getSimpleName()));
        }
    }

    private void renderCode(String code, String url) {
        runOnUiThread(() -> {
            codeView.setText(code.substring(0, 3) + " " + code.substring(3));
            urlView.setText(url);
        });
    }

    private void showStatus(String text) {
        runOnUiThread(() -> statusView.setText(text));
    }

    private void showFailure(String text) {
        Log.w(TAG, text);
        runOnUiThread(() -> {
            statusView.setText(text);
            Toast.makeText(LitetubeActivationActivity.this, text, Toast.LENGTH_LONG).show();
        });
    }

    private void proceedToSplash() {
        runOnUiThread(() -> {
            startActivity(new Intent(LitetubeActivationActivity.this, SplashActivity.class)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK));
            finish();
        });
    }

    @Override
    protected void onDestroy() {
        stopFlag.set(true);
        ioExecutor.shutdownNow();
        super.onDestroy();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_DPAD_CENTER || keyCode == KeyEvent.KEYCODE_ENTER) {
            LitetubePrefs prefs = LitetubePrefs.instance(this);
            String code = prefs.getActivationCode();
            if (code != null) {
                ioExecutor.submit(() -> {
                    JSONObject poll = LitetubeApi.pollDevice(prefs.getLitetubeApiBase(), code);
                    if (poll != null) {
                        String status = poll.optString("status");
                        if ("claimed".equals(status)) {
                            String jwt = poll.optString("jwt");
                            if (jwt != null && !jwt.isEmpty()) {
                                prefs.setJwt(jwt);
                                proceedToSplash();
                            }
                        } else if ("expired".equals(status)) {
                            showFailure("Код истёк. Перезапустите приложение.");
                        }
                    }
                });
                return true;
            }
        }
        return super.onKeyDown(keyCode, event);
    }
}

File: tv-fork-patches/patched/smarttubetv/src/main/java/com/liskovsoft/smartyoutubetv2/tv/ui/litetube/LitetubeActivationActivity.kt
package com.liskovsoft.smartyoutubetv2.tv.ui.litetube

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.qrcode.QRCodeWriter
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import com.liskovsoft.sharedutils.mylogger.Log
import com.liskovsoft.smartyoutubetv2.R
import com.liskovsoft.smartyoutubetv2.common.litetube.LitetubeApi
import com.liskovsoft.smartyoutubetv2.common.litetube.LitetubePrefs
import com.liskovsoft.smartyoutubetv2.tv.ui.main.SplashActivity
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * First-launch activation screen for Litetube TV. Renders the 6-digit code +
 * QR pointing at https://litetube.trfnv.ru/activate?code=XXXXXX, then
 * long-polls /api/devices/poll. Once claimed, persists JWT and forwards
 * upstream SplashActivity to the normal launcher flow.
 *
 * Designed for leanback: title at top, big centered QR + code, instructions
 * below. Single-screen, no scrolling, no focus trap. Confirmed with the
 * D-pad: tap-pairs OK → finished.
 */
class LitetubeActivationActivity : Activity() {

    private val tag = LitetubeActivationActivity::class.java.simpleName
    private val stopFlag = AtomicBoolean(false)
    private val ioExecutor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "litetube-activation").apply { isDaemon = true }
    }

    private lateinit var codeView: TextView
    private lateinit var qrView: ImageView
    private lateinit var statusView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_litetube_activation)
        codeView = findViewById(R.id.litetube_code)
        qrView = findViewById(R.id.litetube_qr)
        statusView = findViewById(R.id.litetube_status)

        val prefs = LitetubePrefs.instance(this)
        val api = prefs.getLitetubeApiBase()
        Log.d(tag, "bootstrapping activation against $api")

        ioExecutor.submit { runActivationLoop(prefs, api) }
    }

    private fun runActivationLoop(prefs: LitetubePrefs, apiBase: String) {
        try {
            val start = LitetubeApi.startDevice(apiBase)
                ?: return@runActivationLoop showFailure("Не удалось связаться с сервером активации. Проверьте интернет.")
            val code = start.optString("code").takeIf { it.length == 6 }
                ?: return@runActivationLoop showFailure("Сервер вернул неожиданный ответ.")
            val qrUrl = start.optString("qr_url")
                .ifBlank { "$apiBase/../activate?code=$code" }
            prefs.setActivationCode(code)
            renderCodeAndQr(code, qrUrl)
            showStatus("Ожидаем подтверждения на телефоне…")

            while (!stopFlag.get()) {
                val poll = LitetubeApi.pollDevice(apiBase, code)
                when (poll?.optString("status")) {
                    "claimed" -> {
                        val jwt = poll.optString("jwt")
                        if (jwt.isBlank()) {
                            showFailure("Сервер не выдал токен. Попробуйте ещё раз.")
                            return@runActivationLoop
                        }
                        prefs.setJwt(jwt)
                        prefs.clearActivationCode()
                        showStatus("Готово. Запускаем плеер…")
                        Thread.sleep(700) // let the user see the success state
                        return@runActivationLoop proceedToSplash()
                    }
                    "expired" -> {
                        showFailure("Код истёк. Перезапустите приложение.")
                        return@runActivationLoop
                    }
                    else -> Thread.sleep(TimeUnit.SECONDS.toMillis(2))
                }
            }
        } catch (t: Throwable) {
            Log.e(tag, "activation loop crashed: ${t.message}", t)
            showFailure("Ошибка: ${t.message ?: t.javaClass.simpleName}")
        }
    }

    private fun renderCodeAndQr(code: String, qrUrl: String) {
        runOnUiThread {
            codeView.text = code.chunked(3).joinToString(" ")
            qrView.setImageBitmap(buildQr(qrUrl, 720, 720))
        }
    }

    private fun showStatus(text: String) {
        runOnUiThread { statusView.text = text }
    }

    private fun showFailure(text: String) {
        Log.w(tag, text)
        runOnUiThread {
            statusView.text = text
            Toast.makeText(this, text, Toast.LENGTH_LONG).show()
        }
    }

    private fun proceedToSplash() {
        runOnUiThread {
            startActivity(Intent(this, SplashActivity::class.java)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK))
            finish()
        }
    }

    /** ZXing is already a transitive dep via chatkit (see settings.gradle). */
    private fun buildQr(content: String, w: Int, h: Int): Bitmap {
        val hints = mapOf<EncodeHintType, Any>(EncodeHintType.ERROR_CORRECTION to ErrorCorrectionLevel.M)
        val matrix = QRCodeWriter().encode(content, BarcodeFormat.QR_CODE, w, h, hints)
        val bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        for (x in 0 until w) {
            for (y in 0 until h) {
                bmp.setPixel(x, y, if (matrix[x, y]) Color.BLACK else Color.WHITE)
            }
        }
        return bmp
    }

    override fun onDestroy() {
        stopFlag.set(true)
        ioExecutor.shutdownNow()
        super.onDestroy()
    }

    /** D-pad ENTER on the QR triggers a manual re-poll — useful if the user
     *  completes the web step while we're between ticks. */
    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if ((keyCode == KeyEvent.KEYCODE_DPAD_CENTER || keyCode == KeyEvent.KEYCODE_ENTER)
            && codeView.visibility == View.VISIBLE) {
            val code = LitetubePrefs.instance(this).getActivationCode() ?: return false
            ioExecutor.submit {
                LitetubeApi.pollDevice(LitetubePrefs.instance(this).getLitetubeApiBase(), code)
                    ?.let(::handlePollResult)
            }
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun handlePollResult(p: org.json.JSONObject) {
        when (p.optString("status")) {
            "claimed" -> {
                val jwt = p.optString("jwt")
                if (jwt.isNotBlank()) {
                    LitetubePrefs.instance(this).setJwt(jwt)
                    proceedToSplash()
                }
            }
            "expired" -> showFailure("Код истёк. Перезапустите приложение.")
        }
    }
}
